### DISCUSSION TITLE
On the idiomatic design of builtin return values

### DISCUSSION THREAD

**Discussion by Xiretza** [fish-shell/fish-shell#9555] [Ideas]
After the discussion in https://github.com/fish-shell/fish-shell/pull/9544#discussion_r1096849586, I decided to write down my thoughts on the idiomatic design of builtins' return values in a more coherent manner. I hope that by detaching this from any particular PR, it'll be easier to discuss things without constantly blocking other more productive work (it can also just be left sitting around until it's "make it pretty" time).

I think there are two related but disctinct problems at play here:

## Return value of builtins' entry functions

Most builtins return exactly one status code, which is used to update `$status`. However, some builtins do not change `$status` at all, thus the current model of `Option<i32>`.

There is one immediate problem with this: the semantics of `Option`'s [`Try`](https://doc.rust-lang.org/core/ops/trait.Try.html#associatedtype.Residual) implementation (the trait backing the `?` operator) do not match this usage. For `Option`, `None` is treated as the error/"residual" value, meaning that if a builtin calls a function returning an `Option` and naively uses `?` on the result, the builtin stops executing and returns "do not change `$status`", which is nonsensical.

The fix would be to either wrap the `Option` in a newtype, or simply create a new enum like this:

```rust
enum StatusAction {
    DontChange,
    Update(i32),
}
```

This type also immediately explains the meaning of builtins' return values to readers.

## Internal functions inside builtins

The story is slightly different for (helper) functions inside builtins. Here, not all status values are created equal: `STATUS_CMD_OK` (value `0`) is special, there are currently 63 occurences of even just the exact string `retval != STATUS_CMD_OK` in `src/builtins/`. The Rust type that most closely matches "success, or one of a set of error values" is `Result`.

A naive approach would be `Result<_, i32>`, where `Ok(_)` represents `STATUS_CMD_OK` and `Err(status)` represents an error. This allows `?` to be used for propagating errors (equivalent to the frequent `if (retval != STATUS_CMD_OK) return retval;` in C++). It can also trivially be converted to an integer status value - `Ok(_)` is `0`, `Err(n)` is `n`.

There is a problem with this though: because the `Err` value can be any status value, and `0` is a valid status value, it is possible to return `Err(0)`, which is ambiguous. According to `?`, it's an error, according to the numeric value, it's success. In order to "make illegal state unrepresentable", the type of the `Err` value needs to change. Luckily, the standard library already provides a handy option here: [`NonZeroI32`](https://doc.rust-lang.org/core/num/struct.NonZeroI32.html), which can be wrapped in a newtype like so:

```rust
#[repr(transparent)]
struct BuiltinError(NonZeroI32);

impl BuiltinError {
    const CMD_ERROR: Self = Self::custom(1);
    const INVALID_ARGS: Self = Self::custom(2);

    pub const fn custom(n: i32) -> Self {
        Self(NonZeroI32::new(n).unwrap())
    }
}
```

It can then be used like so:

```rust
fn foo() -> Result<Thing, BuiltinError> {
    match something()? {
        42 => Err(BuiltinError::CMD_ERROR),
        43 => Err(BuiltinError::custom(130)),
        44 => Ok(something),
    }
}
```

Because `NonZeroI32` has some special attributes, the layout of `Result<(), BuiltinError>` is actually guaranteed to be the same as a plain `i32`: `0` encodes the `Ok(())` variant, the other values encode the errors, making it an equivalent but more meaningful replacement for `i32` status values in general. As such, it can also be used for top-level builtin return values, with the following revised `StatusAction` type:

```rust
enum StatusAction {
    DontChange,
    Update(Result<(), BuiltinError>),
}
```

**Comment by ridiculousfish** [comment]
Thank you for writing this up. I think aiming to support `?` in the way you describe is a worthy goal for the future, and would be pleasant for the reasons you give.

However, for the initial Rust port, idiomatic Rust is an [explicit non-goal](https://github.com/fish-shell/fish-shell/blob/riir/doc_internal/fish-riir-plan.md#approach):

> The Rust code will initially resemble the replaced C++. Fidelity to existing code is more important than Rust idiomaticity, to aid review and bisecting. 

When we find a bug, we can do a nearly line-by-line comparison. Trying to port and simultaneously refactor is riskier than a pure port.

In C++ we use `maybe_t<int>` as the return type of builtins, so in Rust we'll use `Optional<c_int>` for now; once the port is done then I look forwards to the sort of refactoring you describe.
